apps/web/src/lib/tiptap/ipfs-image.tsx

```diff
@@ -0,0 +1,182 @@
+  const formData = new FormData();
+  formData.append("image", file);
+
+  const res = await fetch("/api/editor/upload", {
+    method: "POST",
+    body: formData,
+    headers: {
+      "Accept": "application/json",
+    },
+  });
+
+  if (!res.ok) {
+    throw new Error("Failed to upload image to IPFS");
+  }
+
+  const { uploads } = await res.json();
+
+  const cid = uploads[0].IpfsHash;
+
+  return cid;
+}
+
+export const IpfsImage = (
+  props: CommandExtensionProps<IpfsImageProps>
+) => {
+  const { cid, file, alt, title } = props.node.attrs;
+  const [objectUrl, setObjectUrl] = useState<string | null>(null);
+
+  const fetchImage = useQuery({
+    queryKey: ["ipfs-image", cid],
+    queryFn: async () => {
+      const res = await fetch(`https://gateway.pinata.cloud/ipfs/${cid}`);
+      const imageBlob = await res.blob();
+      return imageBlob;
+    },
+    refetchOnWindowFocus: false,
+    staleTime: 3 * 60 * 1000,
+    cacheTime: 5 * 60 * 1000,
+    enabled: cid !== null,
+  });
+
+  const uploadImage = useMutation({
+    mutationFn: uploadImageToIpfs,
+    onSuccess: (cid) => {
+      props.updateAttributes({
+        cid,
+        file: undefined,
+        alt: file?.name,
+        title: `${file?.name} uploaded to IPFS with CID ${cid}`,
+      });
+    },
+  });
+
+  const isError = uploadImage.isError || fetchImage.isError;
+  const isLoading = uploadImage.isLoading || fetchImage.isLoading || objectUrl === null;
+
+  useEffect(() => {
+    if (file && cid === null && !uploadImage.isLoading) {
+      setObjectUrl(URL.createObjectURL(file));
+      uploadImage.mutate(file);
+    }
+  }, [cid, file, uploadImage]);
+
+  useEffect(() => {
+    if (fetchImage.data) {
+      setObjectUrl(URL.createObjectURL(fetchImage.data));
+    }
+  }, [fetchImage.data]);
+
+  useEffect(() => {
+    return () => {
+      if (objectUrl) {
+        URL.revokeObjectURL(objectUrl);
+      }
+    };
+  }, [objectUrl]);
+
+  if (isError) {
+    return (
+      <NodeViewWrapper as="span">
+        <DataPill query={fetchImage}>no data</DataPill>;
+      </NodeViewWrapper>
+    );
+  }
+
+  return (
+    <NodeViewWrapper as="span">
+      <picture
+        className="relative inline-block drag-handle"
+        data-drag-handle
+        draggable={true}
+        contentEditable={false}
+      >
+        <img
+          src={objectUrl || ""}
+          alt={alt}
+          title={title}
+          className="inline-block"
+          style={{ filter: isLoading ? "blur(4px)" : "none" }}
+        />
+        {isLoading && (
+          <Loader2
+            width={32}
+            height={32}
+            style={{ top: "calc(50% - 16px)", left: "calc(50% - 16px)" }}
+            className="animate-spin absolute"
+          />
+        )}
+      </picture>
+    </NodeViewWrapper>
+  );
+};
+
+export const extension = Node.create({
+  name: 'ipfs-image',
+
+  addOptions() {
+    return {
+      inline: false,
+      HTMLAttributes: {},
+    }
+  },
```

Comment:
Will we ever change this?

Reply:
Most likely not, this was copied from the `@tiptap/extension-image` package.